File: skill/plugin/restart-auth.ts
/**
 * /restart slash command — 5-tier auth fallback (issue #215)
 *
 * Architectural fix shipped 3.3.7-rc.1 after 3.3.6-rc.1 QA found that
 * default-config users (no `commands.ownerAllowFrom`, no
 * `channels.<provider>.allowFrom`) hit "You are not authorized to use
 * this command." when they typed `/restart` to recover from the plugin
 * tool-binding race (the dominant first-run install path).
 *
 * The plugin's `/restart` registration overrides OpenClaw's built-in
 * `/restart` (plugin commands are matched BEFORE built-ins; see
 * upstream `auto-reply/reply/commands-plugin.ts`). With
 * `requireAuth: false` the channel-layer auth check is skipped, and
 * this module's `resolveRestartAuth` decides allow-vs-reject using a
 * five-tier fallback. If the result is `allow`, the caller fires
 * `process.kill(process.pid, 'SIGUSR1')` — which the gateway accepts
 * iff `commands.restart=true` (the default).
 *
 * Tier order (highest priority first):
 *   1. `commands.ownerAllowFrom` explicitly lists invoker → allow
 *   2. `channels.<provider>.allowFrom` explicitly lists invoker → allow
 *   3. Invoker is the same identity this channel session is bound to → allow
 *   4. `credentials.json` exists AND was paired via this same channel → allow
 *   5. BOTH allow-from configs are unset (default) AND only one user
 *      has ever messaged this gateway → allow (lone-user heuristic)
 *
 * Rejection ONLY when explicit config exists and excludes the invoker
 * (i.e. tier 1 or tier 2 was configured but did not match), and no
 * later tier matched.
 *
 * Intentionally pure (no fs/process side effects) so the matrix can be
 * exhaustively tested in `restart-auth.test.ts`. Filesystem / process
 * lookups are passed in via `RestartAuthDeps`.
 */

/** Minimal config shape this module reads. Avoids importing OpenClaw's
 * full OpenClawConfig type so the file stays self-contained for testing. */
export interface RestartAuthConfig {
  commands?: {
    ownerAllowFrom?: ReadonlyArray<string | number>;
    allowFrom?: Record<string, ReadonlyArray<string | number>>;
  };
  channels?: Record<
    string,
    {
      allowFrom?: ReadonlyArray<string | number>;
    } | undefined
  >;
}

/** Verdict returned to the caller. */
export type RestartAuthVerdict =
  | { allow: true; reason: 'tier1-owner-allow-from' | 'tier2-channel-allow-from' | 'tier3-session-bound' | 'tier4-credentials-paired' | 'tier5-lone-user' }
  | { allow: false; reason: 'explicit-deny-owner' | 'explicit-deny-channel' | 'no-tier-matched' };

/** Helper: normalize an allow-from entry for case-insensitive comparison.
 * Accepts string | number (Telegram chat IDs are numeric, Discord uses
 * `discord:<id>` strings, etc.). Mirrors OpenClaw's `normalizeStringEntries`
 * + lowercasing dance — we keep this local so the plugin doesn't depend on
 * OpenClaw's internal symbol layout. */
function normalizeEntry(value: string | number): string {
  return String(value).trim().toLowerCase();
}

function entryMatches(allowFrom: ReadonlyArray<string | number>, senderId: string): boolean {
  if (!senderId) return false;
  const needle = senderId.trim().toLowerCase();
  if (!needle) return false;
  for (const raw of allowFrom) {
    const normalized = normalizeEntry(raw);
    if (!normalized) continue;
    // wildcard: match any sender
    if (normalized === '*') return true;
    if (normalized === needle) return true;
    // Channel-prefixed entries: `telegram:12345`, `discord:user:12345`, etc.
    // The senderId is bare (e.g. `12345`); treat the suffix after the LAST
    // colon as the bare id and compare. This mirrors how OpenClaw parses
    // chat-allow-target prefixes in the upstream allow-from helper.
    const lastColon = normalized.lastIndexOf(':');
    if (lastColon >= 0) {
      const tail = normalized.slice(lastColon + 1);
      if (tail === needle) return true;
    }
  }
  return false;
}

/**
 * Pure dependencies for the auth resolver. Tests inject mocks; production
 * passes real fs / runtime functions.
 *
 * `loadCredentialsExists` returns true if the plugin's own `credentials.json`
 * is present on disk (does NOT load the contents — we never need the
 * mnemonic for this check). The plugin already exposes `loadCredentialsJson`
 * for this; the wrapper just maps `null → false`.
 *
 * `wasPairedViaChannel` is the tier-4 hint: returns true if the active
 * channel is the same one the credentials were paired through. Production
 * persists the pair-channel into the channel-allow-from store at
 * pair-finish (see `pair-session-store.ts` 3.3.0 path) — so the check
 * collapses to "any allow-from store entry exists for this provider".
 *
 * `getDistinctInboundUserCount` returns how many distinct sender IDs have
 * messaged this gateway on the active channel. Tier-5 only fires when this
 * is exactly 1 (the lone-user heuristic — first-run install path).
 */
export interface RestartAuthDeps {
  loadCredentialsExists: () => boolean;
  wasPairedViaChannel: (channel: string) => boolean;
  getDistinctInboundUserCount: (channel: string) => number;
}

export interface RestartAuthInput {
  /** The invoker (Telegram chat id, Discord user id, etc). May be empty if the channel adapter couldn't resolve it. */
  senderId: string | undefined | null;
  /** Channel slug (e.g. `telegram`, `discord`, `slack`). May be undefined for in-process / CLI invocations. */
  channel: string | undefined | null;
  /** OpenClaw config object (the plugin command handler receives this). */
  config: RestartAuthConfig | undefined | null;
}

/**
 * Resolve whether the given invoker may run `/restart`.
 *
 * Tier order: see file header.
 *
 * Edge cases:
 *  - `senderId` empty / undefined → tier 1+2 cannot match (entryMatches
 *    returns false on empty), tier 3 also cannot match. Tier 4 still
 *    works (it's not sender-keyed). Tier 5 still works (it's a count, not
 *    sender-keyed). Default-config + 0 inbound users → 'no-tier-matched'.
 *  - `channel` empty → tier 2/3/4/5 cannot resolve (they're channel-
 *    scoped); only tier 1 can save the day.
 *  - `config` null/undefined → treated as default-config (no allowFrom
 *    set anywhere) → tiers 4 + 5 still apply.
 */
export function resolveRestartAuth(
  input: RestartAuthInput,
  deps: RestartAuthDeps,
): RestartAuthVerdict {
  const senderId = (input.senderId ?? '').toString().trim();
  const channel = (input.channel ?? '').toString().trim().toLowerCase();
  const cfg = input.config ?? {};

  // ---------------------------------------------------------------
  // Tier 1: commands.ownerAllowFrom explicitly lists invoker.
  // ---------------------------------------------------------------
  const ownerAllowFrom = cfg.commands?.ownerAllowFrom;
  const ownerListConfigured = Array.isArray(ownerAllowFrom) && ownerAllowFrom.length > 0;
  if (ownerListConfigured && senderId && entryMatches(ownerAllowFrom!, senderId)) {
    return { allow: true, reason: 'tier1-owner-allow-from' };
  }

  // Note: also honor `commands.allowFrom` per-provider entries as a tier-1
  // equivalent — they are an alternative explicit owner allowlist surface.
  const cmdAllowFromGlobal = cfg.commands?.allowFrom?.['*'];
  const cmdAllowFromChannel = channel ? cfg.commands?.allowFrom?.[channel] : undefined;
  const cmdAllowFromConfigured = Array.isArray(cmdAllowFromGlobal) && cmdAllowFromGlobal.length > 0
    || Array.isArray(cmdAllowFromChannel) && cmdAllowFromChannel.length > 0;
  if (
    senderId
    && (
      (Array.isArray(cmdAllowFromGlobal) && entryMatches(cmdAllowFromGlobal, senderId))
      || (Array.isArray(cmdAllowFromChannel) && entryMatches(cmdAllowFromChannel, senderId))
    )
  ) {
    return { allow: true, reason: 'tier1-owner-allow-from' };
  }

  // ---------------------------------------------------------------
  // Tier 2: channels.<provider>.allowFrom explicitly lists invoker.
  // ---------------------------------------------------------------
  const channelAllowFrom = channel ? cfg.channels?.[channel]?.allowFrom : undefined;
  const channelListConfigured = Array.isArray(channelAllowFrom) && channelAllowFrom.length > 0;
  if (channelListConfigured && senderId && entryMatches(channelAllowFrom!, senderId)) {
    return { allow: true, reason: 'tier2-channel-allow-from' };
  }

  // ---------------------------------------------------------------
  // Tier 3: session-bound identity match.
  //
  // If the channel is paired (i.e. the channel-allow-from store has an
  // entry for this channel + matches this sender), treat that as
  // implicit owner-auth. This covers the case where the user paired
  // via QR earlier in the same install — the pairing wrote a store
  // entry for `<channel>:<senderId>` even though `commands.ownerAllowFrom`
  // is unset.
  //
  // Implementation: `wasPairedViaChannel` returns true if a store
  // entry exists for this channel; we additionally require that the
  // sender's id is one of the store entries (we can't query-by-sender
  // without exposing the store, so we approximate by checking whether
  // the channel has ANY pairing AND there's only one user — that's the
  // common case AND the failure mode the bug report describes; the
  // multi-user-paired case is rare in default-config and falls through
  // to tier 5 if the count is 1, otherwise to no-tier-matched).
  // ---------------------------------------------------------------
  if (channel && senderId && deps.wasPairedViaChannel(channel)) {
    // For tier 3 to fire safely we need either (a) only one paired user
    // on this channel, or (b) the sender is explicitly the paired user.
    // Without exposing the full store list (which would require an
    // upstream API change), we conservatively gate on a single inbound
    // user — same-shape as tier 5, but priority-ordered above it
    // because the pairing is a stronger signal than the lone-user
    // heuristic alone.
    if (deps.getDistinctInboundUserCount(channel) === 1) {
      return { allow: true, reason: 'tier3-session-bound' };
    }
  }

  // ---------------------------------------------------------------
  // Tier 4: credentials.json exists AND paired via this same channel.
  // ---------------------------------------------------------------
  if (channel && deps.loadCredentialsExists() && deps.wasPairedViaChannel(channel)) {
    return { allow: true, reason: 'tier4-credentials-paired' };
  }

  // ---------------------------------------------------------------
  // Tier 5: lone-user heuristic (default config + single inbound user).
  // Only applies when NEITHER explicit allow-from is set anywhere AND
  // exactly one user has messaged this gateway on this channel.
  // ---------------------------------------------------------------
  const anyExplicitConfig = ownerListConfigured || cmdAllowFromConfigured || channelListConfigured;
  if (!anyExplicitConfig && channel && deps.getDistinctInboundUserCount(channel) === 1) {
    return { allow: true, reason: 'tier5-lone-user' };
  }

  // ---------------------------------------------------------------
  // No tier matched — decide WHICH rejection reason to return.
  // ---------------------------------------------------------------
  if (ownerListConfigured || cmdAllowFromConfigured) return { allow: false, reason: 'explicit-deny-owner' };
  if (channelListConfigured) return { allow: false, reason: 'explicit-deny-channel' };
  return { allow: false, reason: 'no-tier-matched' };
}

/** Human-readable rejection text. Kept in this module so the
 * plugin handler doesn't have to duplicate the string. */
export function rejectMessageFor(reason: 'explicit-deny-owner' | 'explicit-deny-channel' | 'no-tier-matched'): string {
  if (reason === 'explicit-deny-owner') {
    return 'You are not authorized to use this command. Add your channel id to `commands.ownerAllowFrom` in your OpenClaw config to grant access.';
  }
  if (reason === 'explicit-deny-channel') {
    return 'You are not authorized to use this command. Add your channel id to `channels.<provider>.allowFrom` in your OpenClaw config to grant access.';
  }
  // 'no-tier-matched' — happens when default-config but multiple inbound
  // users (lone-user heuristic does not apply). Surface a clear pointer.
  return 'You are not authorized to use this command. Multiple users have messaged this gateway; configure `commands.ownerAllowFrom` to identify the owner.';
}

File: python/src/totalreclaw/hermes/restart_auth.py
"""5-tier auth fallback for /restart (issue #215, 2.3.6rc1).

Hermes plugin parity with the OpenClaw plugin's
``skill/plugin/restart-auth.ts``. The matrix is identical so a future
Hermes upstream change can plug this in directly.

NOTE on Hermes-side wiring: as of Hermes ``hermes-agent`` 2026.4.x the
plugin context API does NOT expose ``register_command()`` (the Hermes
team has it on the roadmap — see ``website/docs/guides/build-a-hermes-
plugin.md`` line 240). Until that lands, this module is **shipped as
an exported util** so:

  * The matrix is unit-testable + version-pinned.
  * When Hermes adds ``register_command``, the plugin's
    ``register()`` can wire ``/restart`` in two lines.
  * Documentation surfaces (SKILL.md, hermes-setup.md) can link the
    user to the per-tier reasoning if their config rejects them.

Tier order (highest priority first):

1. ``commands.ownerAllowFrom`` explicitly lists invoker → allow
2. ``channels.<provider>.allow_from`` explicitly lists invoker → allow
3. Invoker is the same identity the channel session is bound to → allow
4. ``credentials.json`` exists AND was paired via this same channel → allow
5. BOTH allow-from configs are unset (default) AND only one user has
   ever messaged this gateway → allow (lone-user heuristic)

Rejection ONLY when explicit config exists and excludes the invoker.

Pure (no fs / network side effects). Filesystem lookups are passed in
via ``RestartAuthDeps`` — the test suite injects mocks; production
plumbing (when Hermes ``register_command`` lands) reads from the
gateway config + the inbound-user tracker.
"""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Any, Callable, Mapping, Optional, Sequence


class RestartAuthVerdict(str, Enum):
    """Reason returned alongside the allow/reject decision.

    Tier 1..5 → allow. ``EXPLICIT_DENY_*`` and ``NO_TIER_MATCHED`` →
    reject. The string values match the OpenClaw plugin's TypeScript
    enum so logs are cross-grep-able.
    """

    TIER1_OWNER_ALLOW_FROM = "tier1-owner-allow-from"
    TIER2_CHANNEL_ALLOW_FROM = "tier2-channel-allow-from"
    TIER3_SESSION_BOUND = "tier3-session-bound"
    TIER4_CREDENTIALS_PAIRED = "tier4-credentials-paired"
    TIER5_LONE_USER = "tier5-lone-user"
    EXPLICIT_DENY_OWNER = "explicit-deny-owner"
    EXPLICIT_DENY_CHANNEL = "explicit-deny-channel"
    NO_TIER_MATCHED = "no-tier-matched"


@dataclass(frozen=True)
class RestartAuthResult:
    allow: bool
    reason: RestartAuthVerdict


@dataclass
class RestartAuthDeps:
    """Filesystem / runtime dependencies. Tests inject mocks; production
    will pass real fs reads (Hermes credentials.json existence) and a
    counter tracker (``InboundUserTracker`` parity TBD on Hermes).
    """

    load_credentials_exists: Callable[[], bool]
    was_paired_via_channel: Callable[[str], bool]
    get_distinct_inbound_user_count: Callable[[str], int]


def _normalize_entry(value: Any) -> str:
    return str(value).strip().lower()


def _entry_matches(allow_from: Sequence[Any], sender_id: str) -> bool:
    if not sender_id:
        return False
    needle = sender_id.strip().lower()
    if not needle:
        return False
    for raw in allow_from:
        normalized = _normalize_entry(raw)
        if not normalized:
            continue
        if normalized == "*":
            return True
        if normalized == needle:
            return True
        # Channel-prefixed entries: `telegram:12345`, `discord:user:12345`, etc.
        # Compare suffix after the LAST colon to support nested prefixes.
        last_colon = normalized.rfind(":")
        if last_colon >= 0:
            tail = normalized[last_colon + 1 :]
            if tail == needle:
                return True
    return False


def resolve_restart_auth(
    sender_id: Optional[str],
    channel: Optional[str],
    config: Optional[Mapping[str, Any]],
    deps: RestartAuthDeps,
) -> RestartAuthResult:
    """Apply the 5-tier fallback. See module docstring for tier order.

    ``config`` is the gateway config dict (Hermes loads this from
    ``~/.hermes/config.yaml``). The relevant sections are
    ``commands.ownerAllowFrom`` / ``commands.allowFrom`` and
    ``channels.<provider>.allow_from`` (Hermes uses snake_case in
    YAML; OpenClaw uses camelCase in JSON5 — the resolver accepts
    both shapes).
    """
    sid = (sender_id or "").strip()
    ch = (channel or "").strip().lower()
    cfg = config or {}

    commands = cfg.get("commands") or {}
    channels = cfg.get("channels") or {}

    # ---------------------------------------------------------------
    # Tier 1: commands.ownerAllowFrom or commands.allowFrom
    # ---------------------------------------------------------------
    owner_allow_from = commands.get("ownerAllowFrom") or commands.get("owner_allow_from")
    owner_list = list(owner_allow_from) if isinstance(owner_allow_from, (list, tuple)) else []
    owner_list_configured = len(owner_list) > 0
    if owner_list_configured and sid and _entry_matches(owner_list, sid):
        return RestartAuthResult(True, RestartAuthVerdict.TIER1_OWNER_ALLOW_FROM)

    cmd_allow_from = commands.get("allowFrom") or commands.get("allow_from") or {}
    cmd_global = cmd_allow_from.get("*") if isinstance(cmd_allow_from, Mapping) else None
    cmd_channel = cmd_allow_from.get(ch) if (isinstance(cmd_allow_from, Mapping) and ch) else None
    cmd_global_list = list(cmd_global) if isinstance(cmd_global, (list, tuple)) else []
    cmd_channel_list = list(cmd_channel) if isinstance(cmd_channel, (list, tuple)) else []
    cmd_allow_configured = len(cmd_global_list) > 0 or len(cmd_channel_list) > 0
    if sid and (
        _entry_matches(cmd_global_list, sid) or _entry_matches(cmd_channel_list, sid)
    ):
        return RestartAuthResult(True, RestartAuthVerdict.TIER1_OWNER_ALLOW_FROM)

    # ---------------------------------------------------------------
    # Tier 2: channels.<provider>.allow_from
    # ---------------------------------------------------------------
    channel_section = channels.get(ch) if ch else None
    channel_allow_from = (
        (channel_section.get("allow_from") if isinstance(channel_section, Mapping) else None)
        or (channel_section.get("allowFrom") if isinstance(channel_section, Mapping) else None)
    )
    channel_list = list(channel_allow_from) if isinstance(channel_allow_from, (list, tuple)) else []
    channel_list_configured = len(channel_list) > 0
    if channel_list_configured and sid and _entry_matches(channel_list, sid):
        return RestartAuthResult(True, RestartAuthVerdict.TIER2_CHANNEL_ALLOW_FROM)

    # ---------------------------------------------------------------
    # Tier 3: session-bound (paired channel + lone inbound user)
    # ---------------------------------------------------------------
    if ch and sid and deps.was_paired_via_channel(ch):
        if deps.get_distinct_inbound_user_count(ch) == 1:
            return RestartAuthResult(True, RestartAuthVerdict.TIER3_SESSION_BOUND)

    # ---------------------------------------------------------------
    # Tier 4: credentials present + paired-via-this-channel
    # ---------------------------------------------------------------
    if ch and deps.load_credentials_exists() and deps.was_paired_via_channel(ch):
        return RestartAuthResult(True, RestartAuthVerdict.TIER4_CREDENTIALS_PAIRED)

    # ---------------------------------------------------------------
    # Tier 5: lone-user heuristic (default config + 1 inbound user)
    # ---------------------------------------------------------------
    any_explicit = owner_list_configured or cmd_allow_configured or channel_list_configured
    if not any_explicit and ch and deps.get_distinct_inbound_user_count(ch) == 1:
        return RestartAuthResult(True, RestartAuthVerdict.TIER5_LONE_USER)

    # ---------------------------------------------------------------
    # No tier matched — pick the most specific rejection reason.
    # ---------------------------------------------------------------
    if owner_list_configured or cmd_allow_configured:
        return RestartAuthResult(False, RestartAuthVerdict.EXPLICIT_DENY_OWNER)
    if channel_list_configured:
        return RestartAuthResult(False, RestartAuthVerdict.EXPLICIT_DENY_CHANNEL)
    return RestartAuthResult(False, RestartAuthVerdict.NO_TIER_MATCHED)


def reject_message_for(reason: RestartAuthVerdict) -> str:
    """Human-readable rejection text — kept here so callers don't have
    to duplicate the strings."""
    if reason == RestartAuthVerdict.EXPLICIT_DENY_OWNER:
        return (
            "You are not authorized to use this command. Add your channel id "
            "to `commands.ownerAllowFrom` in your Hermes config to grant access."
        )
    if reason == RestartAuthVerdict.EXPLICIT_DENY_CHANNEL:
        return (
            "You are not authorized to use this command. Add your channel id "
            "to `channels.<provider>.allow_from` in your Hermes config to grant "
            "access."
        )
    return (
        "You are not authorized to use this command. Multiple users have "
        "messaged this gateway; configure `commands.ownerAllowFrom` to "
        "identify the owner."
    )
